src/parser.rs

```diff
@@ -106,31 +100,57 @@ define_parser!(ParseUnaryOp, UnaryOpKind, |_, state| {
         Ok((state, UnaryOpKind::Negate))
     } else if let Ok((state, _)) = ParseSymbol(Symbol::Hash).parse(state) {
         Ok((state, UnaryOpKind::Length))
-    } else if let Ok((state, _)) = ParseKeyword(Keyword::Not).parse(state) {
+    } else if let Ok((state, _)) = ParseSymbol(Symbol::Not).parse(state) {
         Ok((state, UnaryOpKind::BooleanNot))
     } else {
         Err(ParseAbort::NoMatch)
     }
 });
 
+lazy_static! {
+    static ref BINARY_OPS: HashMap<&'static str, BinaryOpKind> = {
+        let mut map = HashMap::new();
+
+        map.insert(Symbol::Plus.to_str(), BinaryOpKind::Add);
+        map.insert(Symbol::Minus.to_str(), BinaryOpKind::Subtract);
+        map.insert(Symbol::Star.to_str(), BinaryOpKind::Multiply);
+        map.insert(Symbol::Slash.to_str(), BinaryOpKind::Divide);
+        map.insert(Symbol::Caret.to_str(), BinaryOpKind::Exponent);
+        map.insert(Symbol::TwoDots.to_str(), BinaryOpKind::Concat);
+
+        map
+    };
+}
+
 struct ParseBinaryOp;
-define_parser!(ParseBinaryOp, BinaryOpKind, |_, state| {
-    if let Ok((state, _)) = ParseSymbol(Symbol::Plus).parse(state) {
-        Ok((state, BinaryOpKind::Add))
-    } else if let Ok((state, _)) = ParseSymbol(Symbol::Minus).parse(state) {
-        Ok((state, BinaryOpKind::Subtract))
-    } else if let Ok((state, _)) = ParseSymbol(Symbol::Star).parse(state) {
-        Ok((state, BinaryOpKind::Multiply))
-    } else if let Ok((state, _)) = ParseSymbol(Symbol::Slash).parse(state) {
-        Ok((state, BinaryOpKind::Divide))
-    } else if let Ok((state, _)) = ParseSymbol(Symbol::Caret).parse(state) {
-        Ok((state, BinaryOpKind::Exponent))
-    } else if let Ok((state, _)) = ParseSymbol(Symbol::TwoDots).parse(state) {
-        Ok((state, BinaryOpKind::Concat))
-    } else {
-        Err(ParseAbort::NoMatch)
+// Not using define_parser! because it causes some weird lifetime issues
+impl<'state> Parser<'state> for ParseBinaryOp {
+    type Item = BinaryOpKind;
+
+    fn parse(&self, state: ParseState<'state>) -> Result<(ParseState<'state>, Self::Item), ParseAbort> {
+        if let Some(token) = state.peek() {
```

Comment:
We can either do the full destructure in a `match` here or just expand our `if-let` (I think):

```rust
if let Some(Token { kind: TokenKind(symbol) }) = state.peek() {
}
```

Reply:
That's much nicer, yes! Fixed now.